source/common/router/req_header_formatter.cc

```diff
@@ -0,0 +1,99 @@
+#include "common/router/req_header_formatter.h"
+
+#include <cstdint>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "common/http/access_log/access_log_formatter.h"
+#include "common/http/headers.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Router {
+
+HeaderFormatterSharedPtr RequestHeaderParser::parseInternal(const std::string& format) {
+  std::string variable_name;
+  if (format.find("%") == 0) {
+    size_t last_occ_pos = format.rfind("%");
+    if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
+      throw EnvoyException(fmt::format(
+          "Incorrect configuration: {}. Expected the variable to be of format %<variable_name>%",
+          format));
+    }
+    variable_name = format.substr(1, last_occ_pos - 1);
```

Comment:
as per @mattklein123 comment on the ticket, req parameters that are sent by client's already get added to the upstream request as headers (These are one that access log formatter gets it via the `REQ(X?Y):Z%` prefix. A lot of the response headers might not be applicable here (bytes received, duration, etc.) as well. Hence, we might need to provide a set of parameters that can be used as headers to the upstream requests. to begin with, i have started with CLIENT_IP and PROTOCOL that might be something that upstream clusters need it.